CRUD for category

Ayrum1158 opened the pull request:
Added CRUD operations for Category entity

Commit:
WIP
Added non API layer services registered for dependency incejtion

Signed-off-by: Oleksandr Korotetskyi <o.korotetskyi@chisw.com>

Commit:
Switched to Kestrel
Added initial migration

Signed-off-by: Oleksandr Korotetskyi <o.korotetskyi@chisw.com>

Commit:
WIP
Successfully managed to POST category into DB
todo: create service in BLL for thease purposes

Signed-off-by: Oleksandr Korotetskyi <o.korotetskyi@chisw.com>

Commit:
For Category added:
1) category service with Create and Update operations
2) controller API Create (post) and Update (put) operations

Signed-off-by: Oleksandr Korotetskyi <o.korotetskyi@chisw.com>

Commit:
Added Read CRUD API feature

Signed-off-by: Oleksandr Korotetskyi <o.korotetskyi@chisw.com>

Commit:
Added Delete CRUD operation for CategoryController

Signed-off-by: Oleksandr Korotetskyi <o.korotetskyi@chisw.com>

Commit:
Added Swagger
Changed default url to SwaggerUI page
Renamed CategoryController to CategoriesController

Signed-off-by: Oleksandr Korotetskyi <o.korotetskyi@chisw.com>

Commit:
WIP

Signed-off-by: Oleksandr Korotetskyi <o.korotetskyi@chisw.com>

Commit:
Added AutoMapper
Replaced code parts with AutoMapper usage where possible

Signed-off-by: Oleksandr Korotetskyi <o.korotetskyi@chisw.com>

Commit:
Renamed Category POCOs for each layer

Signed-off-by: Oleksandr Korotetskyi <o.korotetskyi@chisw.com>

Commit:
Added Category validation in BLL
Refactored CategoryService a bit

Signed-off-by: Oleksandr Korotetskyi <o.korotetskyi@chisw.com>

solianyk-sergey commented on InternetStoreCHIReloaded/API/Controllers/CategoriesController.cs:
```diff
@@ -0,0 +1,89 @@
+    [ApiController]
+    public class CategoriesController : ControllerBase
+    {
+        private readonly ICategoryService categoryService;
```
@Ayrum1158  According to the conventions, it should be named as `_categoryService`

solianyk-sergey commented on InternetStoreCHIReloaded/API/Controllers/CategoriesController.cs:
```diff
@@ -0,0 +1,89 @@
+    public class CategoriesController : ControllerBase
+    {
+        private readonly ICategoryService categoryService;
+        private readonly IMapper mapper;
```
Please, rename to the `_mapper`

solianyk-sergey commented on InternetStoreCHIReloaded/API/Controllers/CategoriesController.cs:
```diff
@@ -0,0 +1,89 @@
+        {
+            var result = categoryService.GetCategories();
+
+            var response = new ResultContract<IEnumerable<CategoryVM>>() { IsSuccessful = result.IsSuccessful, Message = result.Message };
```
Let's add some `GenericResponse` model. It should exists near other ViewModels. `GenericResponse` should has same properties as `ResultContract`.
We should try to avoid situation when BLL model is returned to the client.

solianyk-sergey commented on InternetStoreCHIReloaded/API/Controllers/CategoriesController.cs:
```diff
@@ -0,0 +1,89 @@
+
+        // GET: api/<CategoryController>
+        [HttpGet]
+        public ResultContract<IEnumerable<CategoryVM>> Get()// return all
```
All API endpoints should be async

solianyk-sergey commented on InternetStoreCHIReloaded/API/Controllers/CategoriesController.cs:
```diff
@@ -0,0 +1,89 @@
+        }
+
+        [HttpPost]
+        public string Post([FromBody] CategoryVM newCategory)// POST aka Create
```
Please, return CategoryVM model instead of string.
Client could be interested in database-generated id or some other properties.

solianyk-sergey replied:
Also, generic response model should be used

solianyk-sergey commented on InternetStoreCHIReloaded/API/Controllers/CategoriesController.cs:
```diff
@@ -0,0 +1,89 @@
+        }
+
+        [HttpPut]// PUT aka Update
+        public ResultContract Put([FromBody] CategoryVM updatedCategory)
```
Please, return also updated CategoryVM

solianyk-sergey replied:
Please, add id parameter

solianyk-sergey commented on InternetStoreCHIReloaded/API/Startup.cs:
```diff
@@ -31,16 +40,43 @@ public void ConfigureServices(IServiceCollection services)
+                .ForMember(dest => dest.CategoryName, opt => opt.MapFrom(src => src.Name))
+                .ForMember(dest => dest.CategoryDescription, opt => opt.MapFrom(src => src.Description))
+                //.ForMember(dest => dest.CreatedDate, opt => opt.MapFrom(src => src.CreatedDate))
+                //.ForMember(dest => dest.UpdatedDate, opt => opt.MapFrom(src => src.UpdatedDate));
```
Please use `.ForMember(dest => dest.CreatedDate, opt => opt.Ignore())` for CreatedDate and UpdatedDate

solianyk-sergey commented on InternetStoreCHIReloaded/API/Startup.cs:
```diff
@@ -31,16 +40,43 @@ public void ConfigureServices(IServiceCollection services)
+            var dbConfig = dbConfigSection.Get<DBConfig>();
+            services.ConfigureDBContext(dbConfig);
+
+            services.AddAutoMapper(cfg =>
```
Please move this configuration to a separate file

Ayrum1158 replied:
Fixed it in next pull request

solianyk-sergey commented on InternetStoreCHIReloaded/API/Startup.cs:
```diff
@@ -31,16 +40,43 @@ public void ConfigureServices(IServiceCollection services)
+
+            services.ConfigureRepositories();// generic and non-generic repositories
+
+            services.AddTransient<ICategoryService, CategoryService>();
```
Please move it to a separate file

solianyk-sergey commented on InternetStoreCHIReloaded/BLL/DependencyInjectionExtensions.cs:
```diff
@@ -0,0 +1,24 @@
+using System.Collections.Generic;
+using System.Text;
+
+namespace BLL
```
Please move it to the API layer. It's not part of the business logic

solianyk-sergey commented on InternetStoreCHIReloaded/BLL/Extensions/CategoryContractExtensions.cs:
```diff
@@ -0,0 +1,23 @@
+
+namespace BLL.Extensions
+{
+    public static class CategoryContractExtensions
```
Looks like this class isn't used

solianyk-sergey commented on InternetStoreCHIReloaded/BLL/Interfaces/ICategoryService.cs:
```diff
@@ -0,0 +1,16 @@
+        public ResultContract DeleteCategory(int id);
+        ResultContract<Category> GetCategory(int id);
+        ResultContract<List<Category>> GetCategories();
+        ResultContract UpdateCategory(Category newCategoryInfo);
```
All methods should return Task<>

solianyk-sergey commented on InternetStoreCHIReloaded/BLL/Services/CategoryService.cs:
```diff
@@ -0,0 +1,182 @@
+{
+    public class CategoryService : ICategoryService
+    {
+        private readonly IGenericRepository<CategoryEntity> categoryRepository;
```
Please rename to `_categoryRepository`

solianyk-sergey commented on InternetStoreCHIReloaded/BLL/Services/CategoryService.cs:
```diff
@@ -0,0 +1,182 @@
+            this.mapper = mapper;
+        }
+
+        private bool Validate(Category category)// only name and description
```
Looks like this method will always say that model is invalid

solianyk-sergey replied:
Please rename to IsValid

solianyk-sergey commented on InternetStoreCHIReloaded/BLL/Services/CategoryService.cs:
```diff
@@ -0,0 +1,182 @@
+
+        private bool Validate(Category category)// only name and description
+        {
+            bool validated = false;
```
Should it be true by default?

solianyk-sergey commented on InternetStoreCHIReloaded/BLL/Services/CategoryService.cs:
```diff
@@ -0,0 +1,182 @@
+            Regex reg;
+
+            reg = new Regex("^[^ ][a-zA-Z ]{3,20}");
+            validated &= reg.IsMatch(category.CategoryName);
```
&& is used frequently than &, it is a bit efficient

solianyk-sergey commented on InternetStoreCHIReloaded/BLL/Services/CategoryService.cs:
```diff
@@ -0,0 +1,182 @@
+            return validated;
+        }
+
+        private bool IsPresentInDB(string categoryName)
```
Please rename to Db

solianyk-sergey commented on InternetStoreCHIReloaded/BLL/Services/CategoryService.cs:
```diff
@@ -0,0 +1,182 @@
+                if (!IsPresentInDB(newCategory.CategoryName))
+                {
+                    var category = mapper.Map<CategoryEntity>(newCategory);
+                    category.CreatedDate = category.UpdatedDate = DateTime.Now;
```
Let's use DateTime.UtcNow

solianyk-sergey commented on InternetStoreCHIReloaded/BLL/Services/CategoryService.cs:
```diff
@@ -0,0 +1,182 @@
+
+                    categoryRepository.Add(category);
+
+                    bool success = categoryRepository.Save() > 0;
```
This is not really great approach to call Add and Save separately. Let's make SaveChanges inside of the Add/Update/Delete methods

solianyk-sergey commented on InternetStoreCHIReloaded/BLL/Services/CategoryService.cs:
```diff
@@ -0,0 +1,182 @@
+            return categoryRepository.FindFirst(c => c.Name == categoryName) != null;
+        }
+
+        public ResultContract AddCategory(Category newCategory)
```
Please, return updated newCategory with Id generated by database

solianyk-sergey commented on InternetStoreCHIReloaded/BLL/Services/CategoryService.cs:
```diff
@@ -0,0 +1,182 @@
+        {
+            categoryRepository.Remove(id);
+
+            bool success = categoryRepository.Save() > 0;
```
Please see comment above. Let's remove Save() method

solianyk-sergey commented on InternetStoreCHIReloaded/BLL/Services/CategoryService.cs:
```diff
@@ -0,0 +1,182 @@
+
+            if (Validate(updatedCategory))
+            {
+                var category = categoryRepository.FindFirst(c => c.Id == updatedCategory.CategoryId);
```
Please rename method to FindFirstOrDefault

solianyk-sergey commented on InternetStoreCHIReloaded/DAL/Migrations/20210723131412_InitialMigration.cs:
```diff
@@ -0,0 +1,33 @@
+                {
+                    Id = table.Column<int>(type: "int", nullable: false)
+                        .Annotation("SqlServer:Identity", "1, 1"),
+                    Name = table.Column<string>(type: "nvarchar(max)", nullable: true),
```
We shouldn't allow max length for most cases. Please add max length constraint

solianyk-sergey commented on InternetStoreCHIReloaded/DAL/Migrations/20210723131412_InitialMigration.cs:
```diff
@@ -0,0 +1,33 @@
+                    Id = table.Column<int>(type: "int", nullable: false)
+                        .Annotation("SqlServer:Identity", "1, 1"),
+                    Name = table.Column<string>(type: "nvarchar(max)", nullable: true),
+                    Description = table.Column<string>(type: "nvarchar(max)", nullable: true),
```
Same as for the Name

solianyk-sergey commented on InternetStoreCHIReloaded/DAL/StoreContext.cs:
```diff
@@ -10,23 +10,23 @@ namespace DAL
             this.dbConfig = dbConfig;
 
-            //Database.Migrate();
+            Database.Migrate();
```
Please remove it from the context constructor.
This method will be called for each new StoreContext object. This could hit performance a lot.

solianyk-sergey reviewed (changes requested):
Please update the code according to the comments

Commit:
Renamed CategoryVM to CategoryViewModel
Added ProductEntity
Added Product seeding
Replaced usage of DateTime.Now with DateTime.UtcNow
Moved DependencyInjectionExtentions to API layer

Signed-off-by: Oleksandr Korotetskyi <o.korotetskyi@chisw.com>

Commit:
Renamed IsPresentInDB to IsPresentInDb in CategoryService
Renamed Validate to IsValid in CategoryService
Fixed IsValid always false output bug
Renamed all private variables according to conventions

Signed-off-by: Oleksandr Korotetskyi <o.korotetskyi@chisw.com>

Commit:
Removed unused class
Reconfigured automapper

Signed-off-by: Oleksandr Korotetskyi <o.korotetskyi@chisw.com>

Commit:
Configured automapper
Replaced some code with automapper usage

Signed-off-by: Oleksandr Korotetskyi <o.korotetskyi@chisw.com>

Commit:
Reworked GenericRepository for better methods understanding
Removed Save method in GenericRepository

Signed-off-by: Oleksandr Korotetskyi <o.korotetskyi@chisw.com>

Commit:
Make CategoryController and CategoryService asynchronous

Signed-off-by: Oleksandr Korotetskyi <o.korotetskyi@chisw.com>

Commit:
Optimised AND comparison in CategoryService

Signed-off-by: Oleksandr Korotetskyi <o.korotetskyi@chisw.com>

Commit:
Fixed unhandled DbUpdateConcurrencyException in CategoryService

Signed-off-by: Oleksandr Korotetskyi <o.korotetskyi@chisw.com>

solianyk-sergey commented on InternetStoreCHIReloaded/BLL/Services/CategoryService.cs:
```diff
@@ -0,0 +1,193 @@
+            return _categoryRepository.FindFirstOrDefault(c => c.Name == categoryName) != null;
+        }
+
+        public async Task<ResultContract> AddCategoryAsync(Category newCategory)
```
Please return added category in the result

solianyk-sergey commented on InternetStoreCHIReloaded/BLL/Services/CategoryService.cs:
```diff
@@ -0,0 +1,193 @@
+
+        private bool IsPresentInDb(string categoryName)
+        {
+            return _categoryRepository.FindFirstOrDefault(c => c.Name == categoryName) != null;
```
Please update repository to be async. Main idea is to have async methods on the repository level. That's why we need to have async services and async controllers.

solianyk-sergey replied:
This change will be done in scope of the next PR

solianyk-sergey commented on InternetStoreCHIReloaded/DAL/Extensions/ModelBuilderExtentions.cs:
```diff
@@ -0,0 +1,71 @@
+            modelBuilder.Entity<CategoryEntity>().HasData(
+                new CategoryEntity()
+                {
+                    Id = -1,
```
Please update ids to start with 1

Ayrum1158 replied:
I watched best practices and it's said that seeding data with negative ids shows us better that it was seeded

solianyk-sergey commented on InternetStoreCHIReloaded/BLL/Services/CategoryService.cs:
```diff
@@ -0,0 +1,193 @@
+            {
+                success = _categoryRepository.Remove(id);
+            }
+            catch(DbUpdateConcurrencyException)// when trying to delete with id not in Db
```
Please move this `try-catch` to the repository level. It's not business-logic responsibility

solianyk-sergey commented on InternetStoreCHIReloaded/BLL/Services/CategoryService.cs:
```diff
@@ -0,0 +1,193 @@
+            catch(DbUpdateConcurrencyException)// when trying to delete with id not in Db
+            {
+                result.IsSuccessful = false;
+                result.Message = $"No entry with Id={id} was found.";
```
It's not necessary to return this message

solianyk-sergey reviewed (changes requested):
Please make changes according to the comments

Commit:
Moved DbUpdateConcurrencyException handling to DAL
Made Categories POST action return newly created category

Signed-off-by: Oleksandr Korotetskyi <o.korotetskyi@chisw.com>

Commit:
Changed seeding data ids to positive

Signed-off-by: Oleksandr Korotetskyi <o.korotetskyi@chisw.com>

Commit:
Relocated DBConfig to DAL

Signed-off-by: Oleksandr Korotetskyi <o.korotetskyi@chisw.com>

solianyk-sergey reviewed (approved):
LGTM (some changes will be done in scope of the next PR)
